python/ray/util/sgd/torch/torch_trainer.py

```diff
@@ -319,24 +303,31 @@ def train(self,
                 logger.info("Retrying training step with %d workers." % len(
                     self.workers))
                 success, worker_stats = self._train_epoch(
-                    num_steps=num_steps, info=info)
+                    num_steps=num_steps, profile=profile, info=info)
         if not success:
             raise RuntimeError("Training run failed.")
 
         worker_stats = ray.get(worker_stats)
+        return self._process_stats(worker_stats)
+
+    def _process_stats(self, worker_stats):
+        stats = {
+            NUM_SAMPLES: sum(
+                stats.pop(NUM_SAMPLES, np.nan) for stats in worker_stats)
+        }
 
-        train_stats = {}
         for stat_key in worker_stats[0]:
             if isinstance(worker_stats[0], numbers.Number):
```

Comment:
Maybe limit this to specific default metrics? e.g. "loss", "accuracy"? The current approach is also fine, but it is a kind of a gotcha if somebody tries to just return a number from training, since the type is the same but the value could be wildly different. We should force the user to think about whether they want just the "master" worker's result, some kind of reduced value, or a list of all values in non-trivial cases instead of assuming something silently. Returning a list of all values is the best default assumption imo since if the user expects a number they will see a type error and be forced to investigate, but with the other two options they might be surprised to see a result different from what they personally returned from the training operator.

Reply:
I think the user's common case is to not want to do any wrangling with the underlying result dictionaries. 

What do you think about a solution where we provide a flag to return all worker stats on `train` (so `dict` per worker)? 

i.e., `train(reduce_results=False)`.

Reply:
`reduce_results=False` sounds good! The fact that this flag will show up in the docs will basically make the current behavior explicit.

Reply:
Merged; just need to add tests. Thanks!